src/statsmodels/contrasts.jl

```diff
@@ -0,0 +1,126 @@
+## Specify contrasts for coding categorical data in model matrix. Contrast types
+## are constructed with data and other options, creating a subtype of
+## AbstractContrast. AbstractContrast provides the interface for creating modle
+## matrix columns and coefficient names
+##
+## ModelFrame will hold a Dict{Symbol, T<:AbstractContrast} that maps column
+## names to contrasts.
+##
+## ModelMatrix will check this dict when evaluating terms, falling back to a
+## default for any categorical data without a specified contrast.
+
+abstract AbstractContrast
+
+termnames(term::Symbol, col::Any, contrast::AbstractContrast) =
+    ["$term - $name" for name in contrast.termnames]
+
+function cols(v::PooledDataVector, contrast::AbstractContrast)
+    ## make sure the levels of the contrast matrix and the categorical data
+    ## are the same by constructing a re-indexing vector. Indexing into
+    ## reindex with v.refs will give the corresponding row number of the
+    ## contrast matrix
+    reindex = [findfirst(contrast.levels, l) for l in levels(v)]
+
+    ## TODO: add kwarg for full-rank contrasts (e.g., when intercept isn't
+    ## specified in a model frame).
+
+    return contrast.matrix[reindex[v.refs], :]
+end
+
+## Default contrast is treatment coding:
+cols(v::PooledDataVector) = cols(v, TreatmentContrast(v))
+termnames(term::Symbol, col::PooledDataArray) = termnames(term, col, TreatmentContrast(col))
+
+
+## Constructing a contrast from a non-pooled data vector will first pool it
+## 
+## Also requires a cols method for non-PooledDataArray column...
+## Base.call{T<: AbstractContrast}(C::Type{T}, v::DataVector, args...; kwargs...) =
+##     Base.call(C, pool(v), args...; kwargs...)
+
+
+## Making a contrast type T only requires that there be a method for
+## contrast_matrix(T, v::PooledDataArray). The rest is boilerplate.
+##
+for contrastType in [:TreatmentContrast, :SumContrast, :HelmertContrast]
+    @eval begin
+        type $contrastType <: AbstractContrast
+            base::Integer
+            matrix::Matrix{Float64}
+            termnames::Vector{Any}
+            levels::Vector{Any}
+        end
+
+        function $contrastType(v::PooledDataVector; base::Integer=1)
+            lvls = levels(v)
+
+            n = length(lvls)
+            n > 1 || error("not enough degrees of freedom to define contrasts")
+            (1 <= base <= n) || error("base = $(base) is not allowed for n = $n")
+
+            not_base = [1:(base-1); (base+1):n]
+            tnames = lvls[not_base]
+
+            mat = contrast_matrix($contrastType, n, base)
+
+            return $contrastType(base, mat, tnames, lvls)
+        end
+    end
+end
+
+## Could write this as a macro, too, so that people can register their own
+## contrast types easily, without having to write out this boilerplate...the
+## downside of that would be that they'd be locked in to a particular set of
+## fields...although they could always just write the boilerplate themselves...
```

Comment:
The boilerplate isn't that bad, I wouldn't bother too much. Creating new contrast types isn't that common AFAIK.